packages/hash/api/src/model/knowledge/linkEntity.model.ts

```diff
@@ -0,0 +1,187 @@
+import {
+  GraphApi,
+  Entity,
+  LinkEntityMetadata,
+  Filter,
+  EntityStructuralQuery,
+} from "@hashintel/hash-graph-client";
+
+import { EntityModel, EntityTypeModel, LinkEntityModel } from "../index";
+
+export type LinkModelConstructorParams = {
+  linkEntity: Entity;
+  linkEntityTypeModel: EntityTypeModel;
+  leftEntityModel: EntityModel;
+  rightEntityModel: EntityModel;
+};
+
+export type LinkModelCreateParams = {
+  ownedById: string;
+  linkEntityTypeModel: EntityTypeModel;
+  leftEntityModel: EntityModel;
+  leftOrder?: number;
+  rightEntityModel: EntityModel;
+  rightOrder?: number;
+  actorId: string;
+};
+
+/**
+ * @class {@link LinkEntityModel}
+ */
+export default class extends EntityModel {
+  leftEntityModel: EntityModel;
+
+  rightEntityModel: EntityModel;
+
+  get linkMetadata(): LinkEntityMetadata {
+    return this.entity.metadata.linkMetadata!;
+  }
+
+  constructor({
+    linkEntity,
+    linkEntityTypeModel,
+    leftEntityModel,
+    rightEntityModel,
+  }: LinkModelConstructorParams) {
+    super({ entity: linkEntity, entityTypeModel: linkEntityTypeModel });
+    this.leftEntityModel = leftEntityModel;
+    this.rightEntityModel = rightEntityModel;
+  }
+
+  static async fromEntity(
+    graphApi: GraphApi,
+    linkEntity: Entity,
+  ): Promise<LinkEntityModel> {
+    if (!linkEntity.metadata.linkMetadata) {
+      throw new Error(
+        `Entity with ID ${linkEntity.metadata.editionId.baseId} is not a link`,
+      );
+    }
+
+    const { entityTypeId } = linkEntity.metadata;
+
+    const [linkEntityTypeModel, leftEntityModel, rightEntityModel] =
+      await Promise.all([
+        EntityTypeModel.get(graphApi, {
+          entityTypeId,
+        }),
+        EntityModel.getLatest(graphApi, {
+          entityId: linkEntity.metadata.linkMetadata.leftEntityId,
+        }),
+        EntityModel.getLatest(graphApi, {
+          entityId: linkEntity.metadata.linkMetadata.rightEntityId,
+        }),
+      ]);
+
+    return new LinkEntityModel({
+      linkEntity,
+      linkEntityTypeModel,
+      leftEntityModel,
+      rightEntityModel,
+    });
+  }
+
+  static async getByQuery(
+    graphApi: GraphApi,
+    filter: Filter,
+    options?: Omit<Partial<EntityStructuralQuery>, "filter">,
+  ): Promise<LinkEntityModel[]> {
+    const { data: subgraph } = await graphApi.getEntitiesByQuery({
+      filter,
+      graphResolveDepths: {
+        dataTypeResolveDepth:
+          options?.graphResolveDepths?.dataTypeResolveDepth ?? 0,
+        propertyTypeResolveDepth:
+          options?.graphResolveDepths?.propertyTypeResolveDepth ?? 0,
+        entityTypeResolveDepth:
+          options?.graphResolveDepths?.entityTypeResolveDepth ?? 0,
+        entityResolveDepth:
+          options?.graphResolveDepths?.entityResolveDepth ?? 0,
+      },
+    });
+
+    return await Promise.all(
```

Comment:
Is this async because `fromEntity` fetches the entity type through the graphAPI?

Reply:
Yes, and the `leftEntityModel`/`rightEntityModel` (to replace the prior `sourceEntityModel`/`targeEntityModel`)